arangod/Aql/OptimizerRules.cpp

```diff
@@ -7889,6 +7905,34 @@ void arangodb::aql::enableAsyncPrefetching(ExecutionPlan& plan) {
   plan.root()->walk(walker);
 }
 
+void arangodb::aql::activateCallstackSplit(ExecutionPlan& plan) {
+  if (willUseV8(plan)) {
+    // V8 requires thread local context configuration, so we cannot 
+    // use our thread based split solution...
+    return;
+  }
+  
+  auto const& options = plan.getAst()->query().queryOptions();
+  struct CallstackSplitter : WalkerWorkerBase<ExecutionNode> {
+    explicit CallstackSplitter(size_t maxNodes) : maxNodesPerCallstack(maxNodes) {}
+    bool before(ExecutionNode* n) override {
+      if (n->getType() == EN::REMOTE) {
```

Comment:
Can we add an assert here that we do not see a Subquery node?
A subquery node is the non-inlined variant that is still present to make Planning easier.
This will branch off and would not account the Subquery itself for the CalStackSize then.
If the Subquery is later inlined we may still break up on the callStack.

e.g.:
```
LET sub = (
  LET q0 = 1
  [...]
  LET qn = 1
  RETURN qn
)
RETURN sub
```
would have a main call stack size of 3 (Singleton, Subquery, Return) before inlining.
While the subquery could be well over the StackSize.

The way this rule is applied I am sure that we do not need to handle the above case, we need to make sure that we not accidentally get into it with refactoring, so an assert is enough.

Reply:
@mchacki: I think in maintainer mode we already walk the plan at the end of all optimizations and make sure that there are no subquery nodes anymore:
https://github.com/arangodb/arangodb/blob/devel/arangod/Aql/Optimizer.cpp#L384

Do you think we should add an extra assertion to the .js test for this here?

Reply:
I would like to have an assertion on the ExecutionNode* n input at exactly this point.

This walker is standalone and is not supporting SubqueryNode.
So I agree right now but outer circumstances it is assured that there will never be a SubqueryNode here, and the assertion will not trigger.
But who knows, in a couple of month someone might move rule ordering around and suddenly we get this Walker at a position where the SubqueryNode is injected and it will silently fail (as it does not react to it)

This assertion is just to save our future selves.

Reply:
Done!